File: src/main/java/core/procedures/continuations/CalledContinuation.java
package core.procedures.continuations;

public class CalledContinuation extends RuntimeException {

  private final Object value;
  private final Continuation continuation;

  CalledContinuation(Object value, Continuation cont) {
    super("CalledContinuationException");
    this.value = value;
    this.continuation = cont;
  }

  public Object getValue() {
    return value;
  }

  public Continuation getContinuation() {
    return continuation;
  }

  /* Do not fill in the execution stack trace (we don't need it anyway) to make CalledContinuations much faster */
  @Override
  public synchronized Throwable fillInStackTrace() {
    return null;
  }

  @Override
  public String toString() {
    return "CalledContinuation{value=" + value + ", continuation=" + continuation + '}';
  }
}

File: src/main/java/core/procedures/continuations/CalledContinuation.kt
package core.procedures.continuations

class CalledContinuation internal constructor(val value: Any, val continuation: Continuation) : RuntimeException("CalledContinuationException") {

    /* Do not fill in the execution stack trace (we don't need it anyway) to make CalledContinuations much faster */
    @Synchronized override fun fillInStackTrace(): Throwable? {
        return null
    }

    override fun toString(): String {
        return "CalledContinuation{value=$value, continuation=$continuation}"
    }
}
